solidity/contracts/vault/TBTCOptimisticMinting.sol

```diff
@@ -0,0 +1,316 @@
+        //       there is no guarantee the wallet will be able to sweep it.
+
+        require(deposit.revealedAt != 0, "The deposit has not been revealed");
+        require(deposit.sweptAt == 0, "The deposit is already swept");
+        require(deposit.vault == address(this), "Unexpected vault address");
+
+        /* solhint-disable-next-line not-rely-on-time */
+        pendingOptimisticMints[depositKey] = block.timestamp;
+
+        emit OptimisticMintingRequested(
+            msg.sender,
+            deposit.depositor,
+            deposit.amount,
+            fundingTxHash,
+            fundingOutputIndex,
+            depositKey
+        );
+    }
+
+    /// @notice Allows a minter to finalize previously requested optimistic
+    ///         minting. The following conditions must be met:
+    ///         - The optimistic minting has been requested for the given
+    ///           deposit.
+    ///         - The deposit has not been swept yet.
+    ///         - At least OPTIMISTIC_MINTING_DELAY passed since the optimistic
+    ///           minting was requested for the given deposit.
+    ///         - The optimistic minting has not been finalized earlier for the
+    ///           given deposit.
+    ///         - The optimistic minting request for the given deposit has not
+    ///           been canceled by a guard.
+    ///         This function mints TBTC and increases pendingOptimisticMints
+    ///         for the given depositor. The finalized optimistic minting
+    ///         request is removed from the contract.
+    function finalizeOptimisticMint(
+        bytes32 fundingTxHash,
+        uint32 fundingOutputIndex
+    ) external onlyMinter {
+        uint256 depositKey = calculateDepositKey(
+            fundingTxHash,
+            fundingOutputIndex
+        );
+
+        uint256 requestedAt = pendingOptimisticMints[depositKey];
+        require(
+            requestedAt != 0,
+            "Optimistic minting not requested or already finalized"
+        );
+        require(
+            /* solhint-disable-next-line not-rely-on-time */
+            block.timestamp - requestedAt > OPTIMISTIC_MINTING_DELAY,
+            "Optimistic minting delay has not passed yet"
+        );
+
+        Deposit.DepositRequest memory deposit = bridge.deposits(depositKey);
+        require(deposit.sweptAt == 0, "The deposit is already swept");
+
+        // TODO: Deal with the minting fee. Right now we mint amount of TBTC
+        //       equal to the output value of the Bitcoin deposit transaction.
+        //       Bridge, when sweeping, will cut a sweeping fee and Bitcoin TX
+        //       fee. If we do not take it into account here, we will create
+        //       an imbalance in the system. Worth noting that even if we do not
+        //       decide to cut fees here, it is possible to fix this imbalance
+        //       later by a donation to the Bridge (see DonationVault).
```

Comment:
There are two cuts on the deposited value the Bridge is doing during the sweeping:
- The deposit treasury fee.
- The Bitcoin sweep TX fee, split equally between all deposits in a sweep.

Handling the first one feels straightforward. We have the fee from the moment of the reveal [stored](https://github.com/keep-network/tbtc-v2/blob/8b9d2629bf4333e650a54f32a4da7cf86bf6785e/solidity/contracts/bridge/Deposit.sol#L94-L95) so we can subtract it from the output value when minting TBTC.

Handling the second one is more complicated. We do not know what the Bitcoin minter fee will be at the moment of sweeping so we can either ignore it or... cut the maximum which [is currently set to 0.001 BTC](https://github.com/keep-network/tbtc-v2/blob/8b9d2629bf4333e650a54f32a4da7cf86bf6785e/solidity/contracts/bridge/Bridge.sol#L300) and has to stay higher than the average to leave some flexibility for the wallet.

For optimistic minting, I'd personally prefer to cut the treasure minting fee but ignore the Bitcoin miner fee. This will create an imbalance in the system but it is the same imbalance as with the move funds transactions between wallets and we have a way to deal with it: an external donation to the Bridge targeted to the [`DonationVault`](https://github.com/keep-network/tbtc-v2/blob/main/solidity/contracts/vault/DonationVault.sol).

Reply:
There is one more thing we could do. Ignore Bitcoin fee as proposed here and in https://github.com/keep-network/tbtc-v2/commit/5ed6239a8a8472d9ae90a94eeb80a066ba12d31f but add an optimistic-minting-specific fee. This fee can be 0 initially but it can also be increased by the Governance later. As long as that fee > the actual average Bitcoin mining fee, the treasury will have a source of Bitcoin to refill the gap. I think that fee is the last missing piece to consider the mechanism from this contract as a candidate to replace the market-based quickminting.

Reply:
Reiterating from our live discussion...

> add an optimistic-minting-specific fee.

yes please! Let's not start it at 0 though, but instead set it to whatever WBTC's current deposit fee is.

Reply:
CoinList charges a flat 0.25% trading fee to wrap or unwrap WBTC (https://coinlist.co/help/what-are-the-fees-to-wrap-and-unwrap-wbtc). If we want to maintain the same fee, for the January release, we have two options:

- If the treasury should receive the fee at the moment of sweeping, set the Deposit fee to 0.25% and the Optimistic Minting fee to 0%.
- If the treasury should receive the fee at the moment of finalizing the Optimistic Mint, set the Optimistic Mint fee to 0.25% and the Deposit Fee to 0%.